#5 feature(Contacts): Выполнена лабораторная работа №5 по ООАиПр

Melaved opened the pull request.

Melaved committed:
feature(Contacts): Проект разделен на 3 части(проекта)
- класс MainVM отнаследован от ObservableObject
- команды теперь реализуются через атрибут RelayCommand из Community MVVM Toolkit
- свойства теперь реализованы через атрибут ObservableProperty
- решение Contacts теперь содержит 3 части(проекта) Contacts.View, Contacts.Model, Contacts.ViewModel

inittose commented on View/App.xaml.cs:
```diff
@@ -2,7 +2,7 @@
 using System.Windows;
 
-namespace View
+namespace Contacts.Model.Services.ViewModel.ViewModel
```
Namespace не соответствует действительности

Melaved replied:
странно, у меня запускается и все работает , но щас все исправлю

Melaved replied:
и не только у себя запускал, все работало

inittose commented on Contacts.Model/Model/Contact.cs:
```diff
@@ -2,7 +2,7 @@
 using System.Text.RegularExpressions;
 
-namespace View.Model
+namespace Contacts.Model.Model
```
Убрать лишнюю папку Model

inittose commented on Contacts.ViewModel/ViewModel/MainVM.cs:
```diff
@@ -1,163 +1,129 @@
+using Contacts.Model.Model.Services;
 
-namespace View.ViewModel
+namespace Contacts.ViewModel.ViewModel
```
Убрать лишнюю папку ViewModel

inittose commented on View/App.xaml.cs:
```diff
@@ -2,7 +2,7 @@
 using System.Data;
```
Убрать неиспользуемые using

inittose commented on Contacts.ViewModel/ViewModel/MainVM.cs:
```diff
@@ -1,163 +1,129 @@
-        /// <param name="parameter">Параметр команды.</param>
         /// <returns>True, если можно применить изменения.</returns>
-        private bool CanApplyChanges(object parameter) => IsEditing && EditingContact != null && !EditingContact.HasErrors;
+        private bool CanApplyChanges() => IsEditing && EditingContact != null && !EditingContact.HasErrors;
```
Строка превышает длину в 100 символов

inittose commented on Contacts.ViewModel/ViewModel/MainVM.cs:
```diff
@@ -1,163 +1,129 @@
 
         /// <summary>
-        /// Возвращает и задает выбранный контакт.
+        /// Вызывается при измении выбранного контакта.
```
*изменении

inittose commented on Contacts.ViewModel/ViewModel/MainVM.cs:
```diff
@@ -1,163 +1,129 @@
+        /// Вызывается при изменении флага редактирования контакта.
         /// </summary>
-        public bool IsEditing
+        /// <param name="value">Новое значение флага редактировани.</param>
```
*редактирования

inittose commented on View/Contacts.View.csproj:
```diff
@@ -14,9 +14,15 @@
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="CommunityToolkit.Mvvm" Version="8.4.0" />
     <PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
```
Эти nuget-пакеты не используются, удалить

inittose commented on Contacts.ViewModel/Contacts.ViewModel.csproj:
```diff
@@ -0,0 +1,18 @@
+
+  <ItemGroup>
+    <PackageReference Include="CommunityToolkit.Mvvm" Version="8.4.0" />
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
```
Не используется

inittose commented on View/MainWindow.xaml:
```diff
@@ -92,23 +90,23 @@
-                    Command="{Binding AddCommand}"
+                    Command="{Binding AddContactCommand}"
                     Content="Add"
                     IsEnabled="{Binding IsEditingContact, Converter={StaticResource InverseBooleanConverter}}"/>
```
Свойства `IsEditingContact` нет в MainVM
IsEnabled у тебя задается через RelayCommand, поэтому удалить строчку

inittose commented on View/Controls/ContactControl.xaml:
```diff
@@ -105,7 +105,7 @@
-            Command="{Binding DataContext.ApplyCommand, RelativeSource={RelativeSource AncestorType=Window}}"
+            Command="{Binding DataContext.ApplyChangesCommand, RelativeSource={RelativeSource AncestorType=Window}}"
             Content="Apply"
             IsEnabled="{Binding DataContext.CanApplyChanges, RelativeSource={RelativeSource AncestorType=Window}}"
```
Свойство `CanApplyChanges` private в MainVM
IsEnabled у тебя задается через RelayCommand, поэтому удалить строчку

inittose reviewed (changes requested):
У меня программа не запускалась. После исправления замечаний ошибка должна уйти. Всегда проверяй работоспособность программы (или фичи) перед коммитом, а тем более перед пулл реквестом

Также реализуй поиск контактов (хоть в методичке об этом не сказано)

Melaved committed:
refactor(Contacts): Исправил зачемания
- удалены неиспользующиеся nuget пакеты
- папка Model удалена
- исправлены неверные пространства имен
- исправлены комментарии
- в XAML для MainWindow и  ContactControl удалены ненужные строки isEnabled(там где это было нужно)

Melaved committed:
feature(Contacts): Добавлена возможность поиска контакта в списке контактов

inittose commented on Contacts.ViewModel/MainVM.cs:
```diff
@@ -1,35 +1,60 @@
+using Contacts.Model.Services;
 
-namespace View.ViewModel
+namespace Contacts.ViewModel.ViewModel
```
Убрать лишний ViewModel у namespace

Melaved replied:
то есть писать Contacts только если во всех проектах решения были изменения 
как вот здесь ?
![image](https://github.com/user-attachments/assets/e408ecdd-7294-4659-b434-c02a835a616e)

inittose replied:
Правильно, ты указываешь, что изменения касаются файлов, которые нельзя отнести к какой-то одной (двух) части программы (или файла)

inittose commented on Contacts.ViewModel/MainVM.cs:
```diff
@@ -95,69 +125,48 @@ public bool IsEditing
-        /// <param name="parameter">Параметр команды.</param>
         /// <returns>True, если можно применить изменения.</returns>
-        private bool CanApplyChanges(object parameter) => IsEditing && EditingContact != null && !EditingContact.HasErrors;
+        private bool CanApplyChanges() => 
```
Лишний пробел в конце строки

inittose commented on Contacts.ViewModel/MainVM.cs:
```diff
@@ -1,35 +1,60 @@
+                {
+                    UpdateFilteredContacts();
+                }
+            };
```
Сделать по аналогии с остальными свойствами (ObservableProperty) через обработчик события OnSearchTextChanged

inittose commented on Contacts.Model/Contact.cs:
```diff
@@ -247,8 +247,8 @@ public Contact Clone()
         {
```
Если ввести в текстовое поле телефона `+7 (999) 123-45-67` как из tooltip, то он не сработает
Поменять либо regex, чтобы можно было записать как в шаблоне, либо поменять шаблон для tooltip

inittose commented on View/App.xaml:
```diff
@@ -1,8 +1,8 @@
 ﻿<Application
-    x:Class="View.App"
+    x:Class="Contacts.Model.Services.ViewModel.ViewModel.App"
```
Неверный путь до класса

inittose commented on View/App.xaml:
```diff
@@ -1,8 +1,8 @@
     xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
     xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
-    xmlns:local="clr-namespace:View"
+    xmlns:local="clr-namespace:Contacts.Model.Services.ViewModel.ViewModel"
```
Неверный путь до namespace

inittose reviewed (changes requested):
Сейчас программа запускается без ошибок
Небольшое замечание по названиям коммитов на будущее: в скобках надо указывать область, где были произведены изменения
Например твой последний коммит на данный момент называется: 
`feature(Contacts): Проект разделен на 3 части(проекта)`
Но изменения были произведены только в файлах MainVM и MainWindow (ViewModel и View соответственно), поэтому в имени коммита стоит указать:
`feature(MainVM, MainWindow): Проект разделен на 3 части(проекта)`
или
`feature(ViewModel, View): Проект разделен на 3 части(проекта)`
Contacts (или sln) можно было бы указать, если измененных файлов в коммите много и они покрывают сразу все части MVVM

Melaved committed:
refactor(Contacts): Исправил замечания
- изменил все неверные пути и пространства имен
- поменял шаблон для tooltip
- сделал поиск через OnSearchTextChanged
